File: openam-mcp-server/src/main/java/org/openidentityplatform/openam/mcp/server/service/UserService.java
/*
 * The contents of this file are subject to the terms of the Common Development and
 * Distribution License (the License). You may not use this file except in compliance with the
 * License.
 *
 * You can obtain a copy of the License at legal/CDDLv1.0.txt. See the License for the
 * specific language governing permission and limitations under the License.
 *
 * When distributing Covered Software, include this CDDL Header Notice in each file and include
 * the License file at legal/CDDLv1.0.txt. If applicable, add the following below the CDDL
 * Header, with the fields enclosed by brackets [] replaced by your own identifying
 * information: "Portions copyright [year] [name of copyright owner]".
 *
 * Copyright 2025 3A Systems LLC.
 */

package org.openidentityplatform.openam.mcp.server.service;

import org.openidentityplatform.openam.mcp.server.config.OpenAMConfig;
import org.openidentityplatform.openam.mcp.server.model.SearchResponseDTO;
import org.openidentityplatform.openam.mcp.server.model.User;
import org.openidentityplatform.openam.mcp.server.model.UserDTO;
import org.springframework.ai.tool.annotation.Tool;
import org.springframework.ai.tool.annotation.ToolParam;
import org.springframework.core.ParameterizedTypeReference;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestClient;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@Service
public class UserService extends OpenAMAbstractService {
    public UserService(RestClient openAMRestClient, OpenAMConfig openAMConfig) {
        super(openAMRestClient, openAMConfig);
    }

    @Tool(name = "get_users", description = "Returns OpenAM user list from the default (root) realm")
    public List<User> getUsers(@ToolParam(required = false, description = "If not set, uses root realm") String realm,
                               @ToolParam(required = false, description = "Username filter") String filter) {

        realm = getRealmOrDefault(realm);

        String queryFilter = "true";
        if(filter != null) {
            queryFilter= "cn sw \"".concat(filter).concat("\"");
        }
        String uri = String.format("/json/realms/%s/users?_queryFilter=%s", realm, queryFilter);
        String tokenId = getTokenId();
        SearchResponseDTO<UserDTO> userSearchResponse  = openAMRestClient.get().uri(uri)
                .header(openAMConfig.tokenHeader(), tokenId)
                .retrieve()
                .body(new ParameterizedTypeReference<>() {});
        return userSearchResponse.result().stream().map(User::new).collect(Collectors.toList());
    }

    private static final Map<String, String> ATTR_MAP = Map.of("familyName", "sn",
            "givenName", "givenName",
            "name", "cn",
            "mail", "mail",
            "phone", "telephoneNumber");

    @Tool(name = "set_user_attribute", description = "Sets the attribute value for a user")
    public User setUserAttribute(@ToolParam(required = false, description = "If not set, uses root realm") String realm,
                                 @ToolParam(description = "username") String username,
                                 @ToolParam(description = "user attribute name") String attribute,
                                 @ToolParam(description = "user attribute value") String value) {

        realm = getRealmOrDefault(realm);

        if(!ATTR_MAP.containsKey(attribute)) {
            throw new RuntimeException(String.format("invalid attribute: %s; allowed values %s", attribute, ATTR_MAP.keySet()));
        }
        String tokenId = getTokenId();

        Map<String, String> requestBody = Map.of(ATTR_MAP.get(attribute), value);
        String uri = String.format("/json/realms/%s/users/%s", realm, username);
        UserDTO user = openAMRestClient.put().uri(uri).body(requestBody)
                .header(openAMConfig.tokenHeader(), tokenId)
                .header("Accept-API-Version","resource=2.0, protocol=1.0")
                .retrieve()
                .body(UserDTO.class);
        return new User(user);
    }


    @Tool(name = "set_user_password", description = "Sets the password for a user")
    public User setUserPassword(@ToolParam(required = false, description = "If not set, uses root realm") String realm,
                                 @ToolParam(description = "username") String username,
                                 @ToolParam(description = "user password") String password) {

        realm = getRealmOrDefault(realm);

        String tokenId = getTokenId();

        Map<String, String> requestBody = Map.of("userpassword", password);
        String uri = String.format("/json/realms/%s/users/%s?_action=changePassword", realm, username);
        UserDTO user = openAMRestClient.put().uri(uri).body(requestBody)
                .header(openAMConfig.tokenHeader(), tokenId)
                .header("Accept-API-Version","resource=2.0, protocol=1.0")
                .retrieve()
                .body(UserDTO.class);
        return new User(user);
    }

    @Tool(name = "create_user", description = "Creates a new user")
    public User createUser(@ToolParam(required = false, description = "If not set, uses root realm") String realm,
                                 @ToolParam(description = "Username (login)") String userName,
                                 @ToolParam(description = "Password (min length 8)") String password,
                                 @ToolParam(required = false, description = "User family name") String familyName,
                                 @ToolParam(required = false, description = "User given name") String givenName,
                                 @ToolParam(required = false, description = "Name") String name,
                                 @ToolParam(required = false, description = "Email") String mail,
                                 @ToolParam(required = false, description = "Phone number") String phone
    ) {

        realm = getRealmOrDefault(realm);

        Map<String, String> userProps = new HashMap<>();
        userProps.put("username", userName);
        userProps.put("userpassword", password);
        if(familyName != null) {
            userProps.put(ATTR_MAP.get("familyName"), familyName);
        }
        if(givenName != null) {
            userProps.put(ATTR_MAP.get("givenName"), givenName);
        }
        if(name != null) {
            userProps.put(ATTR_MAP.get("name"), name);
        }
        if(mail != null) {
            userProps.put(ATTR_MAP.get("mail"), mail);
        }
        if(phone != null) {
            userProps.put(ATTR_MAP.get("phone"), phone);
        }

        String tokenId = getTokenId();

        String uri = String.format("/json/realms/%s/users/?_action=create", realm);
        UserDTO user = openAMRestClient.post().uri(uri).body(userProps)
                .header(openAMConfig.tokenHeader(), tokenId)
                .header("Accept-API-Version","resource=2.0, protocol=1.0")
                .retrieve()
                .body(UserDTO.class);
        return new User(user);
    }

    @Tool(name = "delete_user", description = "Deletes a user")
    public Map<String, String> deleteUser(@ToolParam(required = false, description = "If not set, uses root realm") String realm,
                           @ToolParam(description = "Username (login)") String username) {

        realm = getRealmOrDefault(realm);

        String tokenId = getTokenId();

        String uri = String.format("/json/realms/%s/users/%s", realm, username);
        return openAMRestClient.delete().uri(uri)
                .header(openAMConfig.tokenHeader(), tokenId)
                .header("Accept-API-Version","resource=2.0, protocol=1.0")
                .retrieve()
                .body(new ParameterizedTypeReference<>() {});
    }
}

File: openam-ui/openam-ui-js-sdk/src/lib/userService.ts
/**
 * The contents of this file are subject to the terms of the Common Development and
 * Distribution License (the License). You may not use this file except in compliance with the
 * License.
 *
 * You can obtain a copy of the License at legal/CDDLv1.0.txt. See the License for the
 * specific language governing permission and limitations under the License.
 *
 * When distributing Covered Software, include this CDDL Header Notice in each file and include
 * the License file at legal/CDDLv1.0.txt. If applicable, add the following below the CDDL
 * Header, with the fields enclosed by brackets [] replaced by your own identifying
 * information: "Portions copyright [year] [name of copyright owner]".
 *
 * Copyright 2025 3A Systems LLC.
 */

import type { AuthError, UserAuthData, UserData } from "./types";

class UserService {


  //http://openam.example.org:8080/openam/json/users?_action=idFromSession

  //http://openam.example.org:8080/openam/json/realms/root/users/demo

  private userUrlTemplate: string;
  private usersUrl: string;

  constructor(openamUrl: string) {
    this.userUrlTemplate = openamUrl.concat("/json/realms/{realm}/users/{userId}");
    this.usersUrl = openamUrl.concat("/json/users");
  }


  getUserIdFromSession = async (): Promise<UserAuthData | null> => {
    try {
      const response = await fetch(this.usersUrl.concat("?_action=idFromSession"), {
        method: "POST",
        mode: "cors",
        credentials: "include",
        headers: {
          "Content-Type": "application/json"
        },
      })
      if (!response.ok) {
        return null;
      }
      return await response.json();
    } catch (e) {
      if (import.meta.env.MODE === 'development') {
        console.log("error getting user id from session", e)
        console.log("fallback to demo user")
        return JSON.parse(usersSuccessfulResponse)
      } else {
        console.log("request error ocurred:", e)
      }
      return null;
    }
  }

  getUserData = async (userId: string, realm: string): Promise<UserData> => {
    try {
      const userUrl = this.getUserUrlFromTemplate(userId, realm);

      const response = await fetch(userUrl, {
        method: "GET",
        mode: "cors",
        credentials: "include",
        headers: {
          "Content-Type": "application/json"
        },
      })
      return await response.json();
    } catch (e) {
      if (import.meta.env.MODE === 'development') {
        console.log("error getting user data", e)
        console.log("fallback to demo user data")
        return JSON.parse(testUserData);
      } else {
        throw e
      }
    }
  }

  saveUserData = async (userId: string, realm: string, userData: UserData): Promise<UserData> => {
    const userUrl = this.getUserUrlFromTemplate(userId, realm);

    const dataToUpdate = {
      givenName: userData.givenName,
      sn: userData.sn,
      mail: userData.mail,
      telephoneNumber: userData.telephoneNumber
    }

    try {
      const response = await fetch(userUrl, {
        method: "PUT",
        mode: "cors",
        credentials: "include",
        headers: {
          "Accept-API-Version": "resource=2.0, protocol=1.0",
          "Content-Type": "application/json"
        },
        body: JSON.stringify(dataToUpdate),
      }
      )
      return await response.json();
    }

    catch (e) {
      if (import.meta.env.MODE === 'development') {
        console.log("error getting user data", e)
        console.log("fallback to demo user data")
        return JSON.parse(testUserData);
      } else {
        throw e
      }
    }
  }

  savePassword = async (userId: string, realm: string, password: string): Promise<void> => {
    const userUrl = this.getUserUrlFromTemplate(userId, realm);

    const dataToUpdate = {
      userPassword: password
    }
    try {
      const response = await fetch(userUrl, {
        method: "PUT",
        mode: "cors",
        credentials: "include",
        headers: {
          "Content-Type": "application/json"
        },
        body: JSON.stringify(dataToUpdate),
      }
      )
      if(!response.ok) {
        const data = await response.json() as AuthError
        throw new Error(data.message)
      }
    } catch (e) {
      if (import.meta.env.MODE === 'development') {
        console.log("error saving password, assume its ok", e)
      } else {
        throw e
      }
    }
  }


  private getUserUrlFromTemplate(userId: string, realm: string): string {
    if (!realm || realm === "" || realm === "/") {
      realm = "root";
    }
    return this.userUrlTemplate.replace("{realm}", realm).replace("{userId}", userId);
  }

}



export { UserService }

const usersSuccessfulResponse = `{
    "id": "demo",
    "realm": "/",
    "dn": "id=demo,ou=user,dc=openam,dc=openidentityplatform,dc=org",
    "successURL": "/openam/console",
    "fullLoginURL": "/openam/UI/Login?realm=%2F"
}`

// const userUnauthorizedResponse = `{"code":401,"reason":"Unauthorized","message":"Access Denied"}`

const testUserData = `{
    "username": "demo",
    "realm": "/",
    "uid": [
        "demo"
    ],
    "universalid": [
        "id=demo,ou=user,dc=openam,dc=openidentityplatform,dc=org"
    ],
    "oath2faEnabled": [
        "1"
    ],
    "objectClass": [
        "iplanet-am-managed-person",
        "inetuser",
        "sunFederationManagerDataStore",
        "sunFMSAML2NameIdentifier",
        "devicePrintProfilesContainer",
        "inetorgperson",
        "sunIdentityServerLibertyPPService",
        "iPlanetPreferences",
        "pushDeviceProfilesContainer",
        "iplanet-am-user-service",
        "forgerock-am-dashboard-service",
        "organizationalperson",
        "top",
        "kbaInfoContainer",
        "sunAMAuthAccountLockout",
        "person",
        "oathDeviceProfilesContainer",
        "iplanet-am-auth-configuration-service"
    ],
    "inetUserStatus": [
        "Active"
    ],
    "dn": [
        "uid=demo,ou=people,dc=openam,dc=openidentityplatform,dc=org"
    ],
    "sn": [
        "demo"
    ],
    "cn": [
        "demo"
    ],
    "createTimestamp": [
        "20250805142017Z"
    ],
    "modifyTimestamp": [
        "20250925124445Z"
    ],
    "roles": [
        "ui-self-service-user"
    ]
}`


